rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+        may be included in the *given* `sub_state`.
+
+        Args:
+          sub_state: substate for which we want to lookup the relevent features
+          attributes: if not None, this specifies the list of the attributes of the
+            `Features` that we're interested in (i.e. all other `Features` contained
+            in the relevant messages will be ignored)
+
+        Returns:
+          a dictionary that maps all the (requested) attributes to a list of `Features`
+
+        Raises:
+          - a `ValueError` if any of the lookup table keys that can be generated from
+            the given substate is not contained in the lookup table
+          - a `RuntimeError` if an inconsistency in the lookup table is detected;
+            more precisely, if features for the same attribute are found in two
+            looked up messages
+        """
+        # If we specify a list of attributes, then we want a dict with one entry
+        # for each attribute back - even if the corresponding list of features is empty.
+        features: Dict[Text, List[Features]] = (
+            dict()
+            if attributes is None
+            else {attribute: [] for attribute in attributes}
+        )
+        # get all keys whose values in the lookup table contain features for the
+        # given substate
+        key_attributes = set(sub_state.keys()).intersection(self.KEY_ATTRIBUTES)
+        for key_attribute in key_attributes:
+            key = self._build_key(sub_state={key_attribute: sub_state[key_attribute]})
+            message = self._table.get(key)
+            if not message:
+                raise ValueError(
+                    f"Unknown key {key}. Cannot retrieve features for substate "
+                    f"{sub_state}"
+                )
+            features_from_message = Features.groupby_attribute(
+                message.features, attributes=attributes
+            )
+            for feat_attribute, feat_value in features_from_message.items():
+                existing_values = features.get(feat_attribute)
+                # Note: the follwing if-s are needed because if we specify a list of
+                # attributes then `features_from_message` will contain one entry per
+                # attribute even if the corresponding feature list is empty.
+                if feat_value and existing_values:
+                    raise RuntimeError(
+                        f"Feature for attribute {feat_attribute} has already been "
+                        f"extracted from a different message stored under a key "
+                        f"in {key_attributes} "
+                        f"that is different from {key_attribute}. This means there's a "
+                        f"redundancy in the lookup table."
+                    )
+                if feat_value:
+                    features[feat_attribute] = feat_value
+        return features
+
+    def lookup_message(self, user_text: Text) -> Message:
```

Comment:
Oh this is another good reason to not call it feature lookup something - because I'm actually using the same lookup table for un-featurized messages in the component that extracts training data (in order to prune it) :)